### PR TITLE
implement in-place assignment (avoids heap allocation)

### PR DESCRIPTION
**Pull request by zmedico**

### CODE REVIEW

**Comment by TinyTinni** [comment]
Hi,
thanks for the fix.
I made another version, using realloc (therefore, no new heap allocations, if `iterType` has same size or the previous is was bigger).

I am just unsure about 2 things in my version:
- currently, I assume, that the iterator is trivially assignable. Maybe I need to add the assign operation. On the other hand the assign operation would behave differently, depending on which iterator was hold before. It doesn't sound like a good idea for me.
- constructor/destructor is not called. At the moment, I don't know if it is good or bad. (trivailly copyable requirement)

(link to the patch)
https://gist.github.com/TinyTinni/587d54ce3fa68ebbc3e9e3fb5ac1151d

Btw. is your target capable of C++11?
I wanted to add some nothrow/constexpr/r-value ref./static_asserts etc. maybe next week.

**Comment by zmedico** [comment]
> currently, I assume, that the iterator is trivially assignable. Maybe I need to add the assign operation. On the other hand the assign operation would behave differently, depending on which iterator was hold before. It doesn't sound like a good idea for me.

The assign operation should be perfectly safe as long as getNumber<IterType>() returns the same number for IterType, no? BTW, maybe you'd like to replace the getNumber/vector thing with a template function that returns a pointer to a statically allocated struct? That's what I'm doing [in here](https://github.com/zmedico/cpp_abs_map_poc/blob/master/src/include/map_iter.hpp).

> constructor/destructor is not called. At the moment, I don't know if it is good or bad. (trivailly copyable requirement)

That's too many assumptions for my taste, and I suspect the assignment operator performance is going to be pretty close in practice. But if it suits your needs better then maybe you should go ahead and do that.

> Btw. is your target capable of C++11?
> I wanted to add some nothrow/constexpr/r-value ref./static_asserts etc. maybe next week.

Yeah C++11 is fine by me. Actually I'm not really using AnyIterator directly right now, I've just used it as example code for a ceph patch that I'm working on in [this branch](https://github.com/zmedico/ceph/tree/abs_interval_set).

File: any_iterator.hpp [hunk 1]
```diff
@@ -115,14 +122,23 @@ class any_iterator : std::iterator<std::bidirectional_iterator_tag, T>
     template <typename IterType>
     any_iterator operator=(const IterType& _iter)
     {
+        size_t new_index = getNumber<IterType>();
```
Comment (zmedico) [inline any_iterator.hpp]: Actually this is only safe if T is added to the getNumber parameter list.

Comment (zmedico) [reply any_iterator.hpp]: Anyway, the other code that I'm working on replaces the getNumber/vector thing like this:
```c++
template<class Key, class T>
struct iter_funcs {
  void* (*copy)(const void*);
  void (*assign)(void*, const void*);
  void (*del)(void*);
  bool (*eq)(const void*, const void*);
  void (*inc)(void*);
  void (*dec)(void*);
  std::pair<const Key, T>(*deref)(const void*);
  T(*get_value)(const void*);
  T(*get_key)(const void*);
  void (*set_value)(void*, const T&);
};

template<class Key, class T, class Iter>
iter_funcs<Key,T>* mut_iter_funcs() {
  static iter_funcs<Key,T> funcs{
    wrappers::copy<Iter>,
    wrappers::assign<Iter>,
    wrappers::del<Iter>,
    wrappers::eq<Iter>,
    wrappers::inc<Iter>,
    wrappers::dec<Iter>,
    wrappers::deref<Key,T,Iter>,
    wrappers::get_value<T,Iter>,
    wrappers::get_key<Key,Iter>,
    wrappers::set_value<T,Iter>
  };
  return &funcs;
}
```

Comment (zmedico) [reply any_iterator.hpp]: My code is actually only intended to be used with a relatively small number of well known Iter types, so the assign operation is guaranteed to work in my case.

**Comment by TinyTinni** [comment]
I uploaded the realloc version and a version without std::vector/getNumber. Thanks for that.
This solution doesn't work well, if, as I said, the Iterator is not trivially assignable (doing something different with the assign operator besides simple assigning). But I cannot really remember, when I faced an Iterator with e.g. ref. counting in it.
The advantage is, that there is no heap allocation, if the iterator type changes, but the iterator size remains the same. In my use case, this is usually true.

If your set of interators are small and well known, like you said, and your assign is trivially (e.g. can be expressed with a `memcpy`), you can also get rid of any heap allocation, giving the any_iterator pointer an upper bound buffer bounded by max(sizeof(T...)).

**Comment by zmedico** [comment]
Okay, thank you!

**Pull request closed without merging**